### PR TITLE
[#6596] Add a component to access local camera

### LINKED ISSUE #6596: Add a component to access local camera
### User Story

As a developer, I want a component giving access to local camera and gallery images. This allows developers to access raw image data.

 This will require proper permissions.

### Description

*Type*: Feature

### PR DESCRIPTION
**Pull request by tbenr**
fixes #6596

- [X] `camera/qr-code` event which starts camera to scan qr codes. It manages success scan, user cancel and permission denied.
- [x] `camera/picture` will be the event which should let the user choose to
  - [x] peek a photo from "gallery"
  - [X] shoot a picture with "camera"

extension example:

https://status-im.github.io/pluto/try.html?hash=QmTu4fjZmNEa6PSRpPPz9qoxq24uK9XQnJFZQg3LQ7avv4

https://get.status.im/extension/ipfs@QmTu4fjZmNEa6PSRpPPz9qoxq24uK9XQnJFZQg3LQ7avv4

### CODE REVIEW

**Comment by jeluard** [comment]
@tbenr Great start! Looks nice!
The ultimate test would be to have an image from camera sent to IPFS, then displayed in an image component. All from extension :)

**Comment by tbenr** [comment]
@jeluard we are on the way to reach that! :)

**Comment by tbenr** [comment]
there are 2 things that can be improved:

- intercept "cancel" event when user goes to image gallery to peek a photo. I don't know how to get this feedback from image-picker when the user goes back without actually choose a photo. [HERE](https://github.com/status-im/status-react/blob/bb3b38406a9df664a8062356b886de8436210b3a/src/status_im/extensions/views.cljs#L54)

- add a couple of hardcoded window titles in the translation files. (how can be done? just adding in **translations/en.json** ?

**Review by jeluard — APPROVED** [review]

File: src/status_im/extensions/camera.cljs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ :extensions/camera-picture-taken
+ (fn [cofx [_ base64 {{:keys [on-success]} :data}]]
+   (fx/merge cofx
+             {:dispatch (on-success {:result base64})}
+             (navigation/navigate-back))))
+
+(handlers/register-handler-fx
+ :extensions/camera-picture-selected
+ (fn [cofx [_ base64 {{:keys [on-success]} :data}]]
+   (navigation/navigate-back cofx)))
```
Comment (flexsurfer) [inline src/status_im/extensions/camera.cljs]: why don't we call `on-success` here?

File: src/status_im/extensions/views.cljs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+     [status-bar/status-bar]
+     [toolbar/toolbar nil
+      [toolbar/nav-button (actions/back
+                           #(do
+                              (re-frame/dispatch [:extensions/camera-cancel context])
+                              (re-frame/dispatch [:navigate-back])))]
+      [toolbar/content-title "send picture"]]
+     [camera/camera {:style         {:flex 1}
+                     :aspect        (:fill camera/aspects)
+                 ;    :captureQuality "480p"
```
Comment (flexsurfer) [inline src/status_im/extensions/views.cljs]: leftovers, can we remove them?

File: src/status_im/extensions/views.cljs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+         [icons/icon :icons/camera {:color :white}]]]]]]))
+
+(defn- open-image-picker [context]
+  (react/show-image-picker
+   (fn [image]
+     (let [path (get (js->clj image) "path")]
+       (process-image-and-finish path context)))))
+
+(defn pick-or-take-picture-list-selection [context]
+  {:title   "send picture"
```
Comment (flexsurfer) [inline src/status_im/extensions/views.cljs]: i18n/label

**Review by flexsurfer — APPROVED** [review]

**Comment by jeluard** [comment]
@tbenr When you get a chance, can you comment on @flexsurfer points?

**Comment by tbenr** [comment]
@jeluard @flexsurfer sure, i'm currently away for work. Tomorrow I should have the time to answer and apply changes

File: src/status_im/extensions/camera.cljs (hunk 1 above)
Comment (tbenr) [reply src/status_im/extensions/camera.cljs]: the fx is not used anymore. I forgot removing it

File: src/status_im/extensions/views.cljs (hunk 2 above)
Comment (tbenr) [reply src/status_im/extensions/views.cljs]: 👍

File: src/status_im/extensions/views.cljs (hunk 3 above)
Comment (tbenr) [reply src/status_im/extensions/views.cljs]: @flexsurfer do I just have to add the english translation to **translations/en.json**?

Comment (tbenr) [reply src/status_im/extensions/views.cljs]: timeout: yes assumed :)

**Comment by jeluard** [comment]
@tbenr Build is failing. I can see at least one warning: `Use of undeclared Var status-im.extensions.views/pick-or-take-picture-list-selection at line 43`

**Comment by tbenr** [comment]
@jeluard I'll check ASAP, locally I don't get this error..

**Comment by tbenr** [comment]
@jeluard it seems I fixed it

**Comment by jeluard** [comment]
@tbenr Nice thanks!

**Comment by jeluard** [comment]
@tbenr Do you still have the video you shared the other day? :)

**Comment by tbenr** [comment]
@jeluard mean when i did the live demo to you in google meet? :) it was not a video :)

**Comment by jeluard** [comment]
@tbenr Ah right too bad I wanted to show it off :)

**Comment by jeluard** [comment]
@tbenr Sorry looks like you have to rebase . :(

**Review by goranjovic — APPROVED** [review]

**Comment by tbenr** [comment]
@jeluard ok!

**Comment by jeluard** [comment]
@tbenr Sorry you will have to rebase again :(

**Comment by tbenr** [comment]
ups. done!

**Comment by tbenr** [comment]
ah.. i resolved the conflict directly here via web  but i think something has broken..

**Comment by tbenr** [comment]
letme force it again

**Comment by tbenr** [comment]
@jeluard it is building now..

**Comment by Serhy** [comment]
@tbenr thanks a lot for the extension provided so we could test it!
There are two issues we see (when using https://get.status.im/extension/ipfs@QmTu4fjZmNEa6PSRpPPz9qoxq24uK9XQnJFZQg3LQ7avv4 as extension installed with latest builds made of this PR (https://ci.status.im/job/status-react/job/prs/job/android/view/change-requests/job/PR-7328/lastSuccessfulBuild/artifact/StatusIm-190207-141550-e4d302-pr.apk for Android / https://i.diawi.com/xHhY2D for iOS):

1) Android only specific (with Android 7, Android 6.0.1): when sending `/camerapicture 0` as the message in order to 'Select from gallery'/'Capture' dialogue appear, along with it the 'Error [object Object] is not ISeqable' appears. After tap OK I can proceed to use the app fine. 

screenshot 2019-02-07 at 19 38 38

2) On both iOS and Android, an attempt to 'Capture' image from `/camerapicture` app crashes with `Error: Error: Error: Error: Unknown icon: :icons/camera` error

screenshot 2019-02-07 at 19 41 30

**To reproduce above:**
1) Open Status and create new account
2) Add a new contact to start 1-1 chat with
3) Install `https://get.status.im/extension/ipfs@QmTu4fjZmNEa6PSRpPPz9qoxq24uK9XQnJFZQg3LQ7avv4` extention
4) Navigate to 1-1 chat -> select `/camerapicture` command -> enter 0 -> tap 'Send' button
**Actual result:** See issue #1 described above (note it's soft-error is for Android only, so tap OK to proceed to next step)
5) Tap 'Capture' option from the appeared 'Send picture' pop-up
**Actual result:** See issue #2 described above

Logs with both issues from Android can be found here: 
[LogcatCameraExtention.log](https://github.com/status-im/status-react/files/2842070/CameraExtention.log)

**Comment by tbenr** [comment]
@Serhy checking!

**Comment by tbenr** [comment]
@Serhy I fixed both issues.

**Comment by Serhy** [comment]
@tbenr right, both issues are fixed.
One small issue left: we need to handle cases when user will try to add non-image in `/camerapicture`. When selecting items from Gallery, both iOS and Android sort Videos and Images under the same Gallery folder (other file extensions like .apk, .pdf, .doc which could be found elsewhere - are disabled for selection). However, in case of selecting video files (which are available in Gallery) `/camerapicture` fires an `null is not an object (evaluating 's.cljs$core$IFn$_invoke$arity$1')` error:


Taking as expected behaviour of User Profile image selection: Open Status -> Profile -> tap EDIT -> choose userpic from device Gallery. In a case of selecting video format file we "ignore" it, - i.e. nothing is updated on userpic. 
@jeluard is that right? Should we handle Video files selection from Gallery when using `/camerapicture` in a way we handle user_image selection now? 

### Preconditions
Record video with default device Camera app 

### Reproduction:
* Open Status and create new account
* Add a new contact to start 1-1 chat with
* Install `https://get.status.im/extension/ipfs@QmTu4fjZmNEa6PSRpPPz9qoxq24uK9XQnJFZQg3LQ7avv4` extention
* Navigate to 1-1 chat -> select `/camerapicture` command -> enter 0 -> tap 'Send' button
* Select `Gallery` -> Video file recorded 
 **Actual result:** screenshot 2019-02-08 at 14 47 39

```
02-08 14:23:04.791   492  2442 V MediaPlayerService: Create new media retriever from pid 10141
02-08 14:23:04.971 10141 10196 D skia    : --- SkImageDecoder::Factory returned null
02-08 14:23:04.971 10141 10196 I System.out: 1
02-08 14:23:04.971 10141 10196 D skia    : --- SkImageDecoder::Factory returned null
02-08 14:23:04.981 10141 10195 D ReactNativeJS: DEBUG [status-im.utils.image-processing:27] - Resized error:  Unable to load source image from path
02-08 14:23:04.981  3076 16523 D NfcService: call the applyRouting
02-08 14:23:04.981 10141 10195 D ReactNativeJS: DEBUG [status-im.utils.handlers:34] - Handling re-frame event:  :extensions/camera-error
02-08 14:23:04.981 10141 10195 I ReactNativeJS: PRETTY PRINTED EXCEPTION
02-08 14:23:04.981 10141 10195 I ReactNativeJS: 
02-08 14:23:04.981 10141 10195 I ReactNativeJS: ***
02-08 14:23:04.981 10141 10195 I ReactNativeJS: NAME: "TypeError"
02-08 14:23:04.981 10141 10195 I ReactNativeJS: MESSAGE: null is not an object (evaluating 's.cljs$core$IFn$_invoke$arity$1')
02-08 14:23:04.981 10141 10195 I ReactNativeJS: 
02-08 14:23:04.981 10141 10195 I ReactNativeJS: index.android.bundle:12:7470656
02-08 14:23:04.981 10141 10195 I ReactNativeJS: index.android.bundle:12:7470738
02-08 14:23:04.981 10141 10195 I ReactNativeJS: index.android.bundle:12:1761481
02-08 14:23:04.981 10141 10195 I ReactNativeJS: invoke_interceptor_fn@index.android.bundle:12:1691787
02-08 14:23:04.981 10141 10195 I ReactNativeJS: invoke_interceptors@index.android.bundle:12:1692088
02-08 14:23:04.981 10141 10195 I ReactNativeJS: execute@index.android.bundle:12:1694537
02-08 14:23:04.981 10141 10195 I ReactNativeJS: handle@index.android.bundle:12:1702950
02-08 14:23:04.981 10141 10195 I ReactNativeJS: re_frame$router$IEventQueue$_process_1st_event_in_queue$arity$1@index.android.bundle:12:1730035
02-08 14:23:04.981 10141 10195 I ReactNativeJS: re_frame$router$IEventQueue$_run_queue$arity$1@index.android.bundle:12:1715020
02-08 14:23:04.981 10141 10195 I ReactNativeJS: index.android.bundle:12:1718200
02-08 14:23:04.981 10141 10195 I ReactNativeJS: re_frame$router$IEventQueue$_fsm_trigger$arity$3@index.android.bundle:12:1722411
02-08 14:23:04.981 10141 10195 I ReactNativeJS: index.android.bundle:12:1716322
02-08 14:23:04.981 10141 10195 I ReactNativeJS: index.android.bundle:24:1801
02-08 14:23:04.981 10141 10195 I ReactNativeJS: k@index.android.bundle:24:643
02-08 14:23:04.981 10141 10195 I ReactNativeJS: q@index.android.bundle:24:1007
02-08 14:23:04.981 10141 10195 I ReactNativeJS: callImmediates@index.android.bundle:24:3202
02-08 14:23:04.981 10141 10195 I ReactNativeJS: value@index.android.bundle:19:2904
02-08 14:23:04.981 10141 10195 I ReactNativeJS: index.android.bundle:19:1377
02-08 14:23:04.981 10141 10195 I ReactNativeJS: value@index.android.bundle:19:2621
02-08 14:23:04.981 10141 10195 I ReactNativeJS: value@index.android.bundle:19:1347
02-08 14:23:04.981 10141 10195 I ReactNativeJS: [native code]
02-08 14:23:04.981 10141 10195 I ReactNativeJS: value@index.android.bundle:19:1282
02-08 14:23:04.981 10141 10195 I ReactNativeJS: [native code]
02-08 14:23:04.981 10141 10195 I ReactNativeJS: 
02-08 14:23:04.981 10141 10195 I ReactNativeJS: ***

```

**Comment by jeluard** [comment]
@tbenr @Serhy Good point we want to filter selection so that only images can be picked up somehow

**Comment by tbenr** [comment]
@Serhy @jeluard yes I understand.. I think for now we should we can ignore videos.. they could be too big to be handled in the current implementation..

**Comment by tbenr** [comment]
ok, let me implement some filtering on peek.. or at least ignore\notify to user.

**Comment by jeluard** [comment]
@tbenr Right! Maybe there could be a fixed set of approved mime type that could be overridden by the event user?
Sorry for the added work..

**Comment by tbenr** [comment]
@jeluard @Serhy Now photo picker shows only photos (on profile image too).
I fixed also the error handling which was causing the error.

I have doubts on managing video too, it may open to different things to manage: different camera setup and maybe resizing\encoding...

**Comment by jeluard** [comment]
@tbenr Thanks again ! We can leave video for another bounty I guess .. :D

**Comment by tbenr** [comment]
@jeluard 👍 also 'cause I'd like to concentrate on map component which is driving me crazy :-D

**Comment by Serhy** [comment]
Looks good to me! Thanks a lot, @tbenr for fixes and for your contribution!

**Comment by jeluard** [comment]
:tada:

**Pull request merged**